google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java

```diff
@@ -661,13 +669,209 @@ public Mapper<ListObjectsRequest.Builder> listObjects() {
     }
   }
 
-  static final class Fields extends RpcOptVal<String>
+  static final class Fields extends RpcOptVal<ImmutableSet<NamedField>>
       implements ObjectSourceOpt, ObjectListOpt, BucketSourceOpt, BucketListOpt {
-    private static final long serialVersionUID = -6861404961336468409L;
 
-    private Fields(String val) {
+    /**
+     * Apiary and gRPC have differing handling of where the field selector is evaluated relative to
+     * the request. For Apiary, it's from the root response document; for gRPC it's from the
+     * collection of results.
+     *
+     * <p>In our current case, this means we exclude some fields when we know it is being consumed
+     * for a gRPC message. Unfortunately, we don't know if the constructed Fields instance is for
+     * use with gRPC when it is instantiated so we must define here.
+     */
+    private static final ImmutableSet<String> grpcExcludedFields =
+        ImmutableSet.of("nextPageToken", "prefixes", "selfLink", "mediaLink", "kind", "id");
+
+    private static final long serialVersionUID = -320337719611149532L;
+
+    private Fields(ImmutableSet<NamedField> val) {
       super(StorageRpc.Option.FIELDS, val);
     }
+
+    @Override
+    public Mapper<ImmutableMap.Builder<StorageRpc.Option, Object>> mapper() {
+      return b -> {
+        String collect =
+            val.stream().map(NamedField::getApiaryName).collect(Collectors.joining(","));
+        return b.put(StorageRpc.Option.FIELDS, collect);
+      };
+    }
+
+    @Override
+    public Mapper<GetBucketRequest.Builder> getBucket() {
+      return b -> b.setReadMask(FieldMask.newBuilder().addAllPaths(getPaths()).build());
+    }
+
+    @Override
+    public Mapper<ListBucketsRequest.Builder> listBuckets() {
+      return b -> b.setReadMask(FieldMask.newBuilder().addAllPaths(getPaths()).build());
+    }
+
+    @Override
+    public Mapper<GetObjectRequest.Builder> getObject() {
+      return b -> b.setReadMask(FieldMask.newBuilder().addAllPaths(getPaths()).build());
+    }
+
+    @Override
+    public Mapper<ListObjectsRequest.Builder> listObjects() {
+      return b -> b.setReadMask(FieldMask.newBuilder().addAllPaths(getPaths()).build());
+    }
+
+    @Override
+    public Mapper<ReadObjectRequest.Builder> readObject() {
+      return b -> b.setReadMask(FieldMask.newBuilder().addAllPaths(getPaths()).build());
+    }
+
+    /**
+     * Define a decoder which can clear out any fields which may have not been selected.
+     *
+     * <p>This approach, isn't ideal at the backside after decoding has already taken place.
+     * However, refactoring the whole model pipeline for both json and grpc is going to be a large
+     * change.
+     */
+    Decoder<Blob, Blob> clearBlobFields() {
```

Comment:
nit: can we make this method name more obvious like clearUnselectedBlobFields or something just so it doesn't seem like someone can clear the FieldMask values with it.

Reply:
Will do